src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_resolver.cc

```diff
@@ -790,139 +768,117 @@ RefCountedPtr<LoadBalancingPolicy::Config>
+         priority < discovery_entry.latest_update->priorities.size();
+         ++priority) {
+      const auto& priority_entry =
+          discovery_entry.latest_update->priorities[priority];
+      Json child_policy;
+      if (!discovery_entry.discovery_mechanism->override_child_policy()
+               .empty()) {
+        child_policy =
+            discovery_entry.discovery_mechanism->override_child_policy();
+      } else {
+        const auto& xds_lb_policy = config_->xds_lb_policy().object_value();
+        if (xds_lb_policy.find("ROUND_ROBIN") != xds_lb_policy.end()) {
+          const auto& localities = priority_entry.localities;
+          Json::Object weighted_targets;
+          for (const auto& p : localities) {
+            XdsLocalityName* locality_name = p.first;
+            const auto& locality = p.second;
+            // Construct JSON object containing locality name.
+            Json::Object locality_name_json;
+            if (!locality_name->region().empty()) {
+              locality_name_json["region"] = locality_name->region();
+            }
+            if (!locality_name->zone().empty()) {
+              locality_name_json["zone"] = locality_name->zone();
+            }
+            if (!locality_name->sub_zone().empty()) {
+              locality_name_json["sub_zone"] = locality_name->sub_zone();
+            }
+            // Add weighted target entry.
+            weighted_targets[locality_name->AsHumanReadableString()] =
+                Json::Object{
+                    {"weight", locality.lb_weight},
+                    {"childPolicy",
+                     Json::Array{
+                         Json::Object{
+                             {"round_robin", Json::Object()},
+                         },
+                     }},
+                };
           }
-          // Add weighted target entry.
-          weighted_targets[locality_name->AsHumanReadableString()] =
+          // Construct locality-picking policy.
+          // Start with field from our config and add the "targets" field.
+          child_policy = Json::Array{
               Json::Object{
-                  {"weight", locality.lb_weight},
-                  {"childPolicy",
-                   Json::Array{
-                       Json::Object{
-                           {"round_robin", Json::Object()},
-                       },
+                  {"weighted_target_experimental",
+                   Json::Object{
+                       {"targets", Json::Object()},
                    }},
-              };
+              },
+          };
+          Json::Object& config =
+              *(*child_policy.mutable_array())[0].mutable_object();
+          auto it = config.begin();
+          GPR_ASSERT(it != config.end());
+          (*it->second.mutable_object())["targets"] =
+              std::move(weighted_targets);
+        } else {
+          auto it = xds_lb_policy.find("RING_HASH");
+          GPR_ASSERT(it != xds_lb_policy.end());
```

Comment:
I know it's pre-existing, but does this mean we'll crash if we get a cluster with a new LB policy configured? Or is the client doing validation before we reach here?